src/react/elements.js

```diff
@@ -137,21 +149,55 @@ function createPropsObject(
           }
         }
       }
+    } else if (defaultProps instanceof AbstractObjectValue) {
+      invariant(false, "TODO: we need to eventually support this");
     }
   }
-  if (props instanceof ObjectValue || props instanceof AbstractObjectValue) {
-    // ensure the props is marked as final
+  invariant(props instanceof ObjectValue || props instanceof AbstractObjectValue);
+  if (props instanceof ObjectValue || (props instanceof AbstractObjectValue && !props.values.isTop())) {
```

Comment:
It's used to determine if something is fully abstract and has no values in its ValuesDomain.

Reply:
I understand what it represents technically. I don’t understand why it makes sense here logically.

Reply:
So if we created props, we know them and can mark them as final.

But if we got them from outside then we don’t know anything about them (and thus can’t mark them as final).

Is that accurate? If it is, can’t we move makeFinal to where we create them (since we *know* what they are at that point)? Or is there something else going on here?

Reply:
You're right, we can also reset key and ref if they're not set already because if we didn't have a a safe key/ref it would have thrown above. Resetting makes the abstract object value have a backing template, thus we can then make it final in all cases. Change committed.